File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/barrel/REICompostRecipe.java
package wraith.fabricaeexnihilo.compatibility.rei.barrel;

import me.shedaniel.rei.api.common.entry.EntryIngredient;
import me.shedaniel.rei.api.common.util.EntryIngredients;
import net.minecraft.item.ItemStack;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public record REICompostRecipe(List<ItemStack> inputs, ItemStack output) {

    public REICompostRecipe(List<ItemStack> inputs, ItemStack output) {
        this.inputs = inputs;
        this.output = output;
    }

    public REICompostRecipe(List<ItemStack> inputs) {
        this(inputs, ItemStack.EMPTY);
    }

    public REICompostRecipe(ItemStack output) {
        this(new ArrayList<>(), output);
    }

    public List<ItemStack> getInputs() {
        return inputs;
    }

    public ItemStack getOutput() {
        return output;
    }

    public List<EntryIngredient> reiInputs() {
        return inputs.stream().map(EntryIngredients::of).toList();
    }

    public List<EntryIngredient> reiOutput() {
        return Collections.singletonList(EntryIngredients.of(output));
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/barrel/REICompostRecipe.kt
package exnihilofabrico.compatibility.rei.barrel

import exnihilofabrico.util.asREIEntry
import net.minecraft.item.ItemStack

// Holds a reversed/composite view of the compost recipes; i.e. many -> one
data class REICompostRecipe(val inputs: List<ItemStack> = listOf(), val output: ItemStack = ItemStack.EMPTY) {
    fun reiInputs() = inputs.map { it.asREIEntry() }
    fun reiOutput() = output.asREIEntry()
}